proxy/src/rate_limiter/limit_algorithm.rs

```diff
@@ -159,37 +164,34 @@ impl DynamicLimiter {
     /// Try to acquire a concurrency [Token], waiting for `duration` if there are none available.
     ///
     /// Returns `None` if there are none available after `duration`.
-    pub async fn acquire_timeout(self: &Arc<Self>, duration: Duration) -> Result<Token, Elapsed> {
-        self.acquire_deadline(Instant::now() + duration).await
+    pub async fn acquire_timeout(
+        self: &Arc<Self>,
+        duration: Duration,
+    ) -> Result<Token, ApiLockError> {
+        tokio::time::timeout(duration, self.acquire()).await?
     }
 
     /// Try to acquire a concurrency [Token], waiting until `deadline` if there are none available.
     ///
     /// Returns `None` if there are none available after `deadline`.
-    pub async fn acquire_deadline(self: &Arc<Self>, deadline: Instant) -> Result<Token, Elapsed> {
+    pub async fn acquire(self: &Arc<Self>) -> Result<Token, ApiLockError> {
         if self.config.initial_limit == 0 {
             // If the rate limiter is disabled, we can always acquire a token.
             Ok(Token::disabled())
         } else {
             let mut notified = pin!(self.ready.notified());
             let mut ready = notified.as_mut().enable();
             loop {
```

Comment:
This loop seemed confusing to me on first read - why do we have `if ready { ... }` always leading to `break` with `ready = true` right below it?

_AFAICT_ it would be equivalent to write the function body as

```rust
if self.config.initial_limit == 0 {
    // If the rate limiter is disabled, we can always acquire a token.
    return Ok(Token::disabled());
}

let mut notified = pin!(self.ready.notified());
let ready = notified.as_mut().enable();

if !ready {
    notified.as_mut().await;
}

let mut inner = self.inner.lock();
if inner.take(&self.ready).is_some() {
    Ok(Token::new(self.clone()))
} else {
    Err(ApiLockError::PermitError)
}
```

That said, perhaps changing it is best left for a follow-up. Or maybe I missed something

Reply:
I, actually, read the documentation of `Notified` properly and I think, now it is how it's suppose to be.

The fact that we were notified but failed to acquire permit means, that there was a race condition (e.g. it was notified from releasing permit and from acquiring previous permit at the same time).
In that cases, it's better to wait for the next `notify_one` call. 

I guess, now it is how it is supposed to be, without exhausting CPU and with waiting for other permit.

Reply:
Ah ok I see, so it was changed back to `notified.set(self.ready.notified());` instead of `break Err(...)`, and now the loop is still meaningful?

Reply:
Yep.